### DISCUSSION TITLE
Meeting notes from March 13, 2023

### DISCUSSION THREAD

**Discussion by glopesdev** [open-ephys/bonsai-onix1#8] [General]
# Agenda

* Introductions and overview
* #3
* #1

**Comment by glopesdev** [comment]
## Meeting Notes

### Overview

The current Bonsai API is stable, but does not reflect the Linux-like usage of the low-level liboni drivers.

#### liboni
* Stable context
* Stable usage:
  1. open context
  2. get device table
  3. manipulate registers
  4. start run
  5. grab data
  6. stop run
  7. close context

#### Bonsai.ONIX
* Context initialization and tear-down is handled by a global manager
* Access to hardware during expression tree build-time
  * Hardware errors might propagate even when workflow is just being opened
  * Changing properties in the workflow immediately writes to hardware
  * Changes in hardware are also reflected in the operators (not ideal for persisting settings)

### Proposals

#### Bonsai.Tinkerforge

In this package, a single connection to a host device is also shared to manage state and stream data across many "bricklets" / devices. Creation is handled explicitly through a source. Downstream operators pick up the connection handle and use it to configure themselves and subscribe to data streams. Also allows for hierarchy (i.e. intermediate devices that expose other streams).

Discussion points:
* Could potentially work for ONIX, but not great experience to have to assemble complex workflows manually
* Include workflows or a custom expression builder could be used to allow users to drop-in single macro-nodes
* Subjects can expose access to multiple device streams (or downstream operators)
* Configuration can flow from workflow to hardware almost exclusively:
  * need way to factory reset
  * need easy way to remap device indices
* Need to ensure exclusive access to read / writer / register streams

### Actionables

- [x] Update clroni to include test drivers
- [x] Make small mockup of package with explicit context creation to gain familiary with ONI api.

**Comment by jonnew** [comment]
Small addition to liboni stable usage:

**liboni** 

- Stable usage:
  - open context
  - get device table
  - manipulate registers
  - start run
  - grab data/**manipulate registers**
  - stop run
  - close context
